### PR TITLE
`\r\n` in JSX text should be considered as single line separator

### PR DESCRIPTION
**Pull request by RReverser**
Currently `\r\n` in JSX content is treated as separate line separators, and that results in incorrect line numbers in location properties, and possibility of having broken source maps.

This fix forces them to be treated as single line separator.

### CODE REVIEW

**Comment by RReverser** [comment]
@jeffmo Ping.

**Comment by zpao** [comment]
Jeff is just getting back from vacation but in the process of moving across the country so might be a bit less responsive than usual, sorry about that. I think this makes total sense though. cc @benjamn, @syranide, who can probably also chime in.

**Comment by RReverser** [comment]
@zpao Thanks!

**Comment by syranide** [comment]
:+1: This is the old bug, that seemed to had been resolved and no one was able to reproduce it. Regardless of if you can reproduce that bug, this is the intended behavior. I'm just gonna check so it shouldn't be pushed upstream instead (beware though, older (?) OSX uses just `\r` for newline so need to parse it solo as well).

PS. Also, unless I'm mistaken, you have a bug, `\r\n\r\n` would be seen as 3 lines.

**Comment by syranide** [comment]
Anyway, it shouldn't land upstream because it's fixed in relevant locations there (and in our code).

It's also implemented as:

```
                if (isLineTerminator(ch)) {
                    if (ch === 13 && source.charCodeAt(index) === 10) {
                        ++index;
                    }
                    ++lineNumber;
                    ...
                }
```

...elsewhere in the code, so I think it makes sense to use the same style here.

We could break it out into a shared function, but might be a bad idea as that should ideally be done upstream. So lets not I think.

**Comment by RReverser** [comment]
> This is the old bug, that seemed to had been resolved and no one was able to reproduce it.

Unfortunately, there is still bug - that's why I've sent pull request. You can see it in action on upstream here: http://requirebin.com/?gist=7296e168e755c318609c.

> Regardless of if you can reproduce that bug, this is the intended behavior.

I don't see how this can be intended behavior. Esprima-FB may be used not just by React itself, but by other JSX-tools as well, and they expect to get correct AST with correct locations.

However, you are right <s>regarding incorrect behavior of `\r\n\r\n` even after the fix, as well as</s> (actually works, see below) about using same solutions as elsewhere in a code. Do you want me to fix PR or would you like to do it on your side?

**Comment by syranide** [comment]
@RReverser I don't have push access, so I think, for moving things along, fix it in your PR and benjamn or jeffmo can just have a look at it when they have time and merge it.

**Comment by RReverser** [comment]
@syranide Actually checked again - `\r\n\r\n` works as expected, too. So the only thing I can do is to move `if (isLineTerminator(...` one level up in the sake of code style, not really a bug though (will do in a minute).

**Comment by RReverser** [comment]
Btw, single `\r` works fine, too.

**Comment by RReverser** [comment]
Fixed to be more similar to other places in code. So now just waiting for @benjamn / @jeffmo.

**Comment by RReverser** [comment]
@zpao @benjamn @jeffmo Guys, this is open for about a month while this is a very simple issue - and already resolved. Can you please review & merge this PR?

**Comment by benjamn** [comment]
Turns out I am not a project collaborator. Huh.

**Comment by RReverser** [comment]
@benjamn Oach. Are there any other project collaborators but @zpao and @jeffmo at all?)

File: esprima.js [hunk 1]
```diff
@@ -5647,8 +5647,13 @@ parseYieldExpression: true
             if (ch === '&') {
                 str += scanXJSEntity();
             } else {
-                index++;
+                ++index;
```
Comment (jeffmo) [inline esprima.js:5650]: Let's just leave this as-is (to preserve blame info)

**Comment by jeffmo** [comment]
Looks good, but was there a specific reason for the second commit (the code style fix)?

**Comment by RReverser** [comment]
@jeffmo Not really, just followed this https://github.com/facebook/esprima/pull/25#issuecomment-50541308 @syranide's comment. You can safely omit / squash it with previous one.

**Comment by RReverser** [comment]
Closing as of e44ee1bf836658fd1b6ba653115d38251b3ba00d.

**Pull request closed without merging**